clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -303,7 +303,14 @@ private void throwIfOffsetOutOfRange() throws OffsetOutOfRangeException {
 
             for (PartitionRecords<K, V> part : this.records) {
                 if (!subscriptions.isFetchable(part.partition)) {
-                    log.debug("Ignoring fetched records for {} since it is no longer fetchable", part.partition);
+                    // this can happen when a rebalance happened or a partition consumption paused
+                    // before fetched records are returned to the consumer's poll call
+                    log.debug("Not returning fetched records for partition {} since it is no longer fetchable", part.partition);
+
+                    // we also need to reset the fetch positions to pretend we did not fetch
+                    // this partition in the previous request at all
+                    subscriptions.fetched(part.partition, part.records.get(0).offset());
```

Comment:
Isn't it a little dangerous to assume that the fetch position in the returned record is still relevant?  Suppose we have the following sequence:

1. User calls poll() and we initiate a fetch.
2. User calls commitSync() (or some other blocking operation) and the fetch returns.
3. User calls seek() with a new offset.
4. User calls poll() and now we've written over their seeked offset.

Maybe instead we should reset the fetched position to the consumed position?